metrics/internal/cache/rbd-mirror.go

```diff
@@ -0,0 +1,264 @@
+	InstanceId  string `json:"instance_id"`
+	ClientId    string `json:"client_id"`
+	Hostname    string `json:"hostname"`
+	CephVersion string `json:"ceph_version"`
+	Leader      bool   `json:"leader"`
+	Health      string `json:"health"`
+}
+
+type RBDMirrorImageStatus struct {
+	Name          string                 `json:"name"`
+	GlobalId      string                 `json:"global_id"`
+	State         string                 `json:"state"`
+	Description   string                 `json:"description"`
+	DaemonService RBDMirrorDaemonService `json:"daemon_service"`
+	LastUpdate    string                 `json:"last_update"`
+	PeerSites     []RBDMirrorPeerSite    `json:"peer_sites"`
+}
+
+type RBDMirrorDaemonService struct {
+	ServiceId  string `json:"service_id"`
+	InstanceId string `json:"instance_id"`
+	DaemonId   string `json:"daemon_id"`
+	Hostname   string `json:"hostname"`
+}
+
+type RBDMirrorPeerSite struct {
+	SiteName    string `json:"site_name"`
+	MirrorUuids string `json:"mirror_uuids"`
+	State       string `json:"state"`
+	Description string `json:"description"`
+	LastUpdate  string `json:"last_update"`
+}
+
+type RBDMirrorPeerSiteDescription struct {
+	BytesPerSecond          float64 `json:"bytes_per_second"`
+	BytesPerSnapshot        float64 `json:"bytes_per_snapshot"`
+	LocalSnapshotTimestamp  int64   `json:"local_snapshot_timestamp"`
+	RemoteSnapshotTimestamp int64   `json:"remote_snapshot_timestamp"`
+	ReplayState             string  `json:"replay_state"`
+}
+
+// Cache mirror data for all CephBlockPools with mirroring enabled
+
+var _ cache.Store = &RBDMirrorStore{}
+
+// RBDMirrorStore implements the k8s.io/client-go/tools/cache.Store
+// interface. It stores rbd mirror data.
+type RBDMirrorStore struct {
+	Mutex sync.RWMutex
+	// Store is a map of Pool UID to RBDMirrorPoolStatusVerbose
+	Store           map[types.UID]RBDMirrorPoolStatusVerbose
+	rbdCommandInput rbdCommandInput
+}
+
+func NewRBDMirrorStore() *RBDMirrorStore {
+	return &RBDMirrorStore{
+		Store:           map[types.UID]RBDMirrorPoolStatusVerbose{},
+		rbdCommandInput: rbdCommandInput{},
+	}
+}
+
+func (s *RBDMirrorStore) WithRBDCommandInput(monitor, id, key string) {
+	s.rbdCommandInput.monitor = monitor
+	s.rbdCommandInput.id = id
+	s.rbdCommandInput.key = key
+}
+
+func (s *RBDMirrorStore) Add(obj interface{}) error {
+	o, err := meta.Accessor(obj)
+	if err != nil {
+		return err
+	}
+
+	pool, ok := obj.(*cephv1.CephBlockPool)
+	if !ok {
+		return fmt.Errorf("unexpected object of type %T", obj)
+	}
+
+	if !pool.Spec.Mirroring.Enabled {
+		klog.Infof("skipping rbd mirror status update for pool %s/%s because mirroring is disabled", pool.Namespace, pool.Name)
+		return nil
+	}
+
+	// run RBD command for pool
+	if s.rbdCommandInput.monitor == "" && s.rbdCommandInput.id == "" && s.rbdCommandInput.key == "" {
+		return fmt.Errorf("unable to get RBD mirror data. RBD command input not specified")
+	}
+
+	klog.Infof("%+v", s.rbdCommandInput)
```

Comment:
is this required?

Reply:
It's helpful but it leaks creds. Will remove it.